Functions, function types, and function calls (#2875)
carbon-language/carbon-lang: proposals/p2875.md

```
fn F[T:! type](x: T) -> T { return x; }
```

introduces a value named `F`, whose type is a unique _function type_. Distinct
functions have distinct function types, even if they have the same signature. A
function type is an empty, trivial type. There is no way to name a function type
other than asking for the type of the function value.

```
// Compile-time function.
fn TypeOf[T:! type](x: T) -> type { return T; }

// ✅ `F` is a first-class value with a first-class type.
alias FType = TypeOf(F);
var my_f: FType = F;
```

Function values are regular values that can be stored in variables, passed to
functions, and so on.

```
fn G() -> i32 {
  // ✅ `my_f` has function type `FType`. This is a direct call to `F`.
  return my_f(1);
}

fn Sort[template F:! type, T:! type](v: Vector(T)*, f: F);
fn Compare(a: i32, b: i32) -> Ordering { return a.(Ordered.Compare)(b); }
fn SortInts(v: Vector(i32)*) { Sort(v, Compare); }
```

For the purpose of the orphan rule, a function type is considered to be declared
by the function declaration that introduces the function value.

### Call syntax

Calls take the form `a(b, c, d)`. This is syntactically equivalent to a primary

> expression followed by a tuple literal, except that no trailing comma is
> required for the case of a single argument, `a(b)`.

Reviewer: ```suggestion
expression followed by a tuple literal, except that the trailing comma is
optional for the case of a single argument, `a(b)`.
```

I'm pretty sure this is the intent? Because what you're saying is "no trailing comma is required", which could be read as saying `a(b,)` is invalid because it has a trailing comma, while the lack thereof is required. However, explorer currently allows the optional trailing comma.

Author: Yeah, I suppose using "no X is required" to mean "X is not required" rather than "X is required to be absent" may be hard for non-native readers to parse. Though I'd prefer to keep "required" here because I think that better contrasts with the rule for tuple literals:

```suggestion
expression followed by a tuple literal, except that a trailing comma is
not required for the case of a single argument, `a(b)`.
```

The callee can be a name, a literal, a member access, or some more complex
expression enclosed parentheses.

### Direct calls

When the callee in a call expression can be resolved to a declaration, either
because it is the name of a parameterized entity like a generic class or
interface, or because it has a function type, the call is a _direct call_.

In a direct call, a call signature is available which is used to check the given
arguments against the callee's declared implicit and explicit parameters. This
checking proceeds as follows:

-   Argument deduction is performed by comparing the declared parameter types
    against the actual parameter types and deducing values for implicit

>     arguments that make the types equal.

Reviewer: Isn't the deduction of `self` a little different, since the object isn't an actual parameter type? Or is that just not covered here since classes defines behavior?

Author: The way I'm viewing this is that `self` is bound when forming the bound member function: in
```
class A { fn F[self: Self, T:! type](x: T); }
fn DoThing(a: A) { a.F(0); }
```
... `a.F` already has a value for `self`, leaving behind a function with signature `[T:! type](x: T)`, which is informally what's [described in the design for member access](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/member_access.md#instance-binding).

-   Then, for each binding in the explicit parameter list in turn, all argument
    values that have been deduced are substituted into the parameter.
    -   If the parameter is a `template :!` binding, the argument expression is
        converted to a constant value phase expression of the type of the
        binding.

>     -   If the parameter is a non-`template` `:!` binding, the argument

Reviewer: ```suggestion
    -   If the parameter is a `:!` binding that is not a `template`, the argument
```

I think this is saying the same as you meant, just suggesting the swap so that the difference of a space not being in backticks isn't significant. This seems particularly subtle for html rendering, but I'm particularly suggesting it because I want to confirm I'm interpreting correctly.

Author: Done, with a different phrasing (I didn't like "that is not a `template`" for bindings because I don't think we want to say that a template parameter is a template).

        expression is converted to a symbolic value phase expression of the type
        of the binding.
    -   Otherwise, the parameter is pattern-matched against the argument.
-   If a parameter is a `:!` binding, its corresponding converted argument
    expression is evaluated, and its value is added to the list of deduced

>     argument values before any later parameters are processed.

Reviewer: My impression is that this actually happens as part of what's stated in 146 because that says "for each binding". Is that accurate? It's just a little confusing because that bullet seems to be saying it does something for each argument, then later you have "but this happens in the middle" without explicitly saying that it's actually during the previous bullet. Maybe it'd help to describe this more as a flow?

- For each binding:
  - Substitute:
    - the three cases
  - If a parameter is a `:!`

Author: Hopefully addressed by removing the bullet symbol from this paragraph. Is this clear enough that the "for each" scopes over the whole bullet, or would it still be better to explicitly indent this within the "for each binding"?

Reviewer: This looks good, thanks!

functions have distinct function types, even if they have the same signature. A
function type is an empty, trivial type. There is no way to name a function type
other than asking for the type of the function value.

```
// Compile-time function.
fn TypeOf[T:! type](x: T) -> type { return T; }

// ✅ `F` is a first-class value with a first-class type.
alias FType = TypeOf(F);
var my_f: FType = F;
```

Function values are regular values that can be stored in variables, passed to
functions, and so on.

```
fn G() -> i32 {
  // ✅ `my_f` has function type `FType`. This is a direct call to `F`.
  return my_f(1);
}

fn Sort[template F:! type, T:! type](v: Vector(T)*, f: F);
fn Compare(a: i32, b: i32) -> Ordering { return a.(Ordered.Compare)(b); }
fn SortInts(v: Vector(i32)*) { Sort(v, Compare); }
```

For the purpose of the orphan rule, a function type is considered to be declared
by the function declaration that introduces the function value.

### Call syntax

Calls take the form `a(b, c, d)`. This is syntactically equivalent to a primary
expression followed by a tuple literal, except that a trailing comma is not

> required for the case of a single argument, `a(b)`.

Reviewer: ```suggestion
required for the case of a single argument, either `a(b)` or `a(b,)`.
```

Just looking for a different way to emphasize the meaning here (to your point about tuple syntax, most of my concern is that this'll be copied into the design, and the direct comparison won't be obvious to a new reader).

Author: I've taken another pass over this section; I think this new approach sufficiently emphasizes what I'm going for here and should still be clear when copied into the design.

    -   [Signature-based function types](#signature-based-function-types)
    -   [Make direct and indirect calls behave uniformly](#make-direct-and-indirect-calls-behave-uniformly)
-   [Future work](#future-work)
    -   [Overloading](#overloading)
    -   [Lambdas](#lambdas)
    -   [Function pointers](#function-pointers)

<!-- tocstop -->

## Abstract

Specify the behavior of function calls and the type and behavior of the entity
introduced by a function declaration.

## Problem

Functions are a primitive building block through which we define Carbon's
execution semantics, but we don't currently have a specification for what a
function is in Carbon nor how function calls work. We have open questions around
whether Carbon has first-class function types, how types can overload the
function call operation, and how callable entities such as the name of a
parameterized class fit into the picture -- are they functions or something
else?

## Background

This proposal assumes the reader understands how functions and function pointers
work in C or C++.

## Proposal

A function declaration in Carbon introduces a new, unique type, called a

> _function type_. The function name is a stateless value of that function type.

Reviewer 2: This was confusing to me because I think of names as _referring to_ values, rather than _being_ values.

```suggestion
A function declaration in Carbon introduces a new, unique type, called a
_function type_. Uses of the function name refer to a stateless value of that function type.
```

Author: Done.

> _function type_. The function name is a stateless value of that function type.

Reviewer 2: In some ways this compromises our "types are values" philosophy, because we're introducing both a type and a singular value of that type, in order to represent the same information in the type and value domains. I don't see how we can avoid that, though, so this is just an observation; no action is needed.

Reviewer 3: @mconst Happening again. Note the argument for this approach in the "alternatives considered" below.

Author: I think any time a type is stateless, we're in some sense representing the same information in the type and value domains, simply because by definition there is no information in the value domain (and we view the type as being part of the value). But note that a bound member function is not stateless -- `x.F` holds `x` in the value domain -- and the same will be true for lambda expressions, so I think in this case function types being stateless is just a degenerate case of a more general pattern of potentially-stateful function-like objects.

-   Values of dependent types that are constrained to be callable.
-   User-defined class types that overload function call syntax.

In the first four cases, argument deduction is used to determine the values of
deduced arguments, and then pattern matching is used to bind parameter values to
arguments.

In the remaining cases, the built-in `Call` interface is used to determine the
semantics of the call, and the call `F(args)` is translated into
`F.(Call(ArgTypes).Op)(args)`. Note that this is itself a bound member function
call, and `Call(ArgTypes)` is a call to a parameterized entity, so this
transformation is only performed once.

## Details

### Functions and function types

A function declaration such as

```
fn F[T:! type](x: T) -> T { return x; }
```

introduces a value named `F`, whose type is a unique _function type_. Distinct
functions have distinct function types, even if they have the same signature. A
function type is an empty, trivial type. There is no way to name a function type
other than asking for the type of the function value.

```
// Compile-time function.
fn TypeOf[T:! type](x: T) -> type { return T; }

// ✅ `F` is a first-class value with a first-class type.

> alias FType = TypeOf(F);

Reviewer 3: I think you can only have names on the right side of an alias, not an expression like this.
```suggestion
let template FType:! type = TypeOf(F);
```

Author: We definitely want to allow `Vector(i32)` on the right hand side of an alias, so it's not just names. But without an actual design for aliases, it's hard to say -- and I like your suggestion for that reason :)

var my_f: FType = F;
```

Function values are regular values that can be stored in variables, passed to
functions, and so on.

```
fn G() -> i32 {
  // ✅ `my_f` has function type `FType`. This is a direct call to `F`.
  return my_f(1);
}

fn Sort[template F:! type, T:! type](v: Vector(T)*, f: F);
fn Compare(a: i32, b: i32) -> Ordering { return a.(Ordered.Compare)(b); }
fn SortInts(v: Vector(i32)*) { Sort(v, Compare); }
```

> For the purpose of the orphan rule, a function type is considered to be declared

Reviewer 2: ```suggestion
For the purpose of the [orphan rule](/docs/design/generics/terminology.md), a function type is considered to be declared
```

Reviewer 3: I'd link to either https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#orphan-rule or https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/terminology.md#coherence

Reviewer 2: Oh, yeah, I meant to include `#coherence` in the suggestion.

Author: Done.

by the function declaration that introduces the function value.

### Call syntax

Calls take the form `a(b, c, d)` or `a(b, c, d,)`, where:

-   `a` is the callee, which can be a name, a literal, a member access, or some
    more complex expression enclosed parentheses.
-   `b`, `c`, `d` are any number of argument expressions. Arguments are
    separated by commas, and if the argument list is not empty, an optional
    trailing comma is permitted but not required after the final argument.

Call syntax is syntactically equivalent to a primary expression followed by a
tuple literal, except that a tuple literal requires a trailing comma to form a
single-element tuple `(b,)`, whereas in call syntax both `a(b)` and `a(b,)` are
permitted.

### Direct calls

> When the callee in a call expression can be resolved to a declaration, either

Reviewer 2: According to the examples below, `cmp((*v)[i], (*v)[j])` isn't a direct call, but `cmp` can be resolved to a declaration. Should this say "function or parameterized entity declaration"?

Author: Rephrased.

    -   If the parameter is a `template :!` binding, the argument expression is
        converted to a constant value phase expression of the type of the
        binding.
    -   If the parameter is a checked `:!` binding, the argument expression is
        converted to a symbolic value phase expression of the type of the
        binding.
    -   Otherwise, the parameter is pattern-matched against the argument.

    If a parameter is a `:!` binding, its corresponding converted argument
    expression is evaluated, and its value is added to the list of deduced
    argument values before any later parameters are processed.

The result of the call expression depends on the callee:

-   If the callee is a parameterized entity such as a generic class or a generic
    interface, the result is the specific instance of that generic, such as a
    class or interface, and the call is a value expression of type `type`.
-   If the callee is a function value, the call is an initializing expression
    whose type is the substituted return type of the function. When evaluated,
    the call expression will invoke the function and produce whatever value it
    returns.

> 

Reviewer 3: Do we want to say anything about functions called from a type position being evaluated at compile time, like `TypeOf` above?

Author: Yes, but not here; I'm writing a separate proposal for that :)

### Generic callable parameters

A generic parameter can be constrained to be a callable type using the `Call`
interface:

```
interface Call(Args: type) {
  let Result:! type;
  fn Op[self: Self](args: Args) -> Result;
}
```

> **TODO:** `Call` should be variadic. For now, we model it as taking a tuple
> type, and we model `Call.Op` as taking a tuple value.

For example:

```
fn Sort[T:! type, F:! Call((T, T)) where .Result = Ordering]
       (v: Vector(T)*, cmp: F) {
```

A call expression that is not a direct call is translated into an invocation of
`Call(Args).Op`, where `Args` is the type of the call's argument tuple.

```
  // In Sort...
  auto ord: auto = cmp((*v)[i], (*v)[j]);
  // ... is translated into ...
  auto ord: auto = cmp.(Call((T, T)).Op)((*v)[i], (*v)[j]);
```

### Function types and `Call`

A `fn` declaration that declares a function with a runtime signature (see below)

> implicitly provides an `impl` of `Call` matching the function signature.

Reviewer 2: ```suggestion
A `fn` declaration implicitly provides an `impl` of `Call` matching the function signature
(unless the function it declares doesn't have a runtime signature; see below).
```

(Rephrasing to make it clearer that you don't need to understand "runtime signature" on a first read)

Author: Rephrased substantially, and removed the term "runtime signature" entirely.

```
fn Select[T:! type](b: bool, x: T, y: T) -> T {
  return if b then x else y;
}

// Generated:
impl forall [T:! type, BType:! ImplicitAs(bool)]

>     Select as (BType, T, T) where .Result = T {

Reviewer: ```suggestion
    Select as Call((BType, T, T)) where .Result = T {
```

Shouldn't the interface name be here?

> example because an argument or return value is not copyable. If we add a
> mechanism to allow an `impl` to specify a different expression category for
> its parameters or result than the one in the interface, for example by taking
> a `var` parameter where the interface took a `let` parameter, that would
> automatically also be available for overloaded function calls.

A _runtime signature_ is one that doesn't depend on any compile-time parameters
in its explicit argument list, nor on any refutable pattern matching between the
call arguments and the function parameters.

```
fn Runtime[T:! type](x: T);
fn CompileTime(T:! type, x: T);
// 🤷 Undecided whether this is valid.
fn RefutablePattern(1 as i32);

fn Run() {
  // ✅ Calls `Runtime(0)`.
  Runtime.(Call(i32).Op)(0);
  // ❌ Can't call `CompileTime` this way, it doesn't have a runtime signature.
  CompileTime.(Call(type, i32).Op)(i32, 0);
  // ❌ Can't call `RefutablePattern` this way, it can't implement `Call(i32)`
  // for arbitrary i32 arguments.
  RefutablePattern.(Call(i32).Op)(0);
}
```

> **Note:** There is no support for passing compile-time parameters to a checked

> > generic function parameter.

Reviewer 2: Does this limitation apply only to indirect calls, or is even something like this invalid?

```
fn f(T:! type);

fn g(T:! type) {
  f(T);
}
```

Author: What I meant was, there is no support for passing a compile-time argument to a compile-time parameter of a value whose type is a checked generic type parameter that is constrained to be `Call`able.

```
fn f(T:! type);
// Error, but would be OK with `template C:! type`.
fn g[C:! Call(type)](c:! C) { c(i32); }
```

But I think saying that is redundant and confusing, so I've just removed this entirely.

### Overloaded call operator

The `Call` interface can be implemented to overload the meaning of the function
call operator for a type.

```
class Func(Arg:! type) {
  impl as Call((Arg,)) where .Result = () {
    fn Op[self: Self](arg: (Arg,)) { Print("hello, world"); }
  }
}

fn Run() {
  let f: Func(i32) = {};
  // ✅ Prints "hello, world".

>   f();

Reviewer 3: Shouldn't this take an `i32` argument?

Author: Done.

}
```

There are no constraints on the callee type, beyond the normal constraints for
implementing an interface.

```
class X { var n: i32; }
// ✅ OK, but inadvisable.
impl {.a: X} as Call(()) where .Result = i32 {
  fn Op[self: Self](args: ()) -> i32 {
    return self.a.n;
  }
}
fn Run() -> i32 {

>   return {.a = {.n = 1} as X}();

Reviewer 3: ```suggestion
  // Returns 1.
  return {.a = {.n = 1} as X}();
```

explicit parameters -- that will be checked against the provided arguments. In
this proposal, that means that a set of overloaded functions introduces a single
function type that supports being called in different ways.

Depending on how we approach overloading, overload selection might be done by
checking each candidate in turn until we find one that matches, or checking them
all and somehow selecting the best match, and our current intent is to use the
former approach. Translated into this proposal, one important consequence is
that the function type corresponding to the set of overloaded functions has
multiple parameterized implementations of the `Call(...)` interface, and the
implementation selected for a particular call will need to follow whatever rules
the overloading mechanism picks. For example, if function overloading picks the
first matching function, then given the following, using placeholder syntax:

```
overloaded fn Abs[T:! Unsigned](x: T) -> T { return x; }
overloaded fn Abs[T:! Floating](x: T) -> T { return x < 0 ? -x : x }
```

we would synthesize implementations that behave as follows:

```
match_first {

>   impl forall [T:! Unsigned] Abs as Call(T) where .Result = T { ... }
>   impl forall [T:! Floating] Abs as Call(T) where .Result = T { ... }

Reviewer: ```suggestion
  impl forall [T:! Unsigned] Abs as Call((T,)) where .Result = T { ... }
  impl forall [T:! Floating] Abs as Call((T,)) where .Result = T { ... }
```

In the current design you've chosen to use tuples instead of variadics, so this should be passing a one-length tuple, right?

}
```

For a type that is `Unsigned & Floating`, the former `impl` will be selected,
matching the behavior of overload selection.

### Lambdas

A lambda is expected to behave much like a function under this proposal, with
the primary difference being that lambdas can be stateful, whereas functions are
stateless.

### Function pointers

For interoperability with C++, and for purposes of cheaply storing references to
stateless functions, function pointers should be supported. A function pointer
could be modeled as:

```

> alias FunctionPtr(Args:! type, Result:! type) =

Reviewer 3: I don't think `alias` can do this.

Author: We don't have a design for `alias` yet, so I don't think we know that one way or the other :) If `alias` can't do this  then we're missing functionality compared to C++ alias templates.

In any case, this is "future work".

else?

## Background

This proposal assumes the reader understands how functions and function pointers
work in C or C++.

## Proposal

A function declaration in Carbon introduces a new, unique, stateless type,
called a _function type_. The function name is bound to a value of that function
type.

Function calls use C-like syntax: an expression naming a callable is followed by
a syntax resembling a tuple of arguments. There are several kinds of callable:

-   Functions, and more generally
    [values of function types](#functions-and-function-types).
-   [Bound methods](#bound-methods), such as `my_vector.Begin`.
-   Lambdas, once they are introduced to Carbon.
-   Parameterized entities, such as a generic class `Vector` or a generic
    interface `AddWith`.
-   Values of dependent types that are
    [constrained to be callable](#generic-callable-parameters).
-   User-defined class types that
    [overload function call syntax](#overloaded-call-operator).

In the first four cases, argument deduction is used to determine the values of
deduced arguments, and then pattern matching is used to bind parameter values to

> arguments.

Reviewer 4: Hmm, I would have expected this to read:

```suggestion
deduced arguments, and then pattern matching is used to bind parameter names to
argument values.
```

WDYT? Or am I missing something important about the prior phrasing?

One possible tweak is that not all parameters are named, so maybe:
```suggestion
deduced arguments, and then pattern matching is used to bind parameters to
argument values.
```

Or:
```suggestion
deduced arguments, and then pattern matching is used to bind parameter patterns to
argument values.
```

```
class HasMember {
  // `HasMember.F` has a stateless function type, with signature
  // `[self: Self](n: i32) -> i32`.
  fn F[self: Self](n: i32) -> i32;
}
fn F(h1: HasMember, h2: HasMember) -> i32 {
  // ✅ `h1.F` is a bound method value whose type is a bound method type,
  // with signature `(n: i32) -> i32`.
  var hf: auto = h1.F;
  // ✅ `h1.F` and `h2.F` are of the same bound method type.
  hf = h2.F;
  // ✅ Same as `h2.F(4)`.
  return hf(4);
}
```

### Call syntax

Calls take the form `a(b, c, d)` or `a(b, c, d,)`, where:

-   `a` is the callee, which can be a name, a literal, a member access, or some
    more complex expression enclosed parentheses.
-   `b`, `c`, `d` are any number of argument expressions. Arguments are
    separated by commas, and if the argument list is not empty, an optional
    trailing comma is permitted but not required after the final argument.

Call syntax is syntactically equivalent to a primary expression followed by a
tuple literal, except that a tuple literal requires a trailing comma to form a
single-element tuple `(b,)`, whereas in call syntax both `a(b)` and `a(b,)` are
permitted.

### Direct calls

> A call expression is a _direct call_ when the callee is:
> 
> -    the name of a parameterized entity, like a generic class or interface,
> -    has a function type or bound method type.

Author: ```suggestion
A call expression is a _direct call_ when the callee:

-   is the name of a parameterized entity, like a generic class or interface, or
-   has a function type or bound method type.
```

In a direct call, a call signature is available which is used to check the given
arguments against the callee's declared implicit and explicit parameters. This
checking proceeds as follows:

-   Argument deduction is performed by comparing the declared parameter types

>     against the actual parameter types and deducing values for implicit

Reviewer 4: Should this be:

```suggestion
    against the actual argument types and deducing values for implicit
```

Not sure though...

    -   Otherwise, the parameter is pattern-matched against the argument.

    If a parameter is a `:!` binding, its corresponding converted argument
    expression is evaluated, and its value is added to the list of deduced
    argument values before any later parameters are processed.

The result of the call expression depends on the callee:

-   If the callee is a parameterized entity such as a generic class or a generic
    interface, the result is the specific instance of that generic, such as a
    class or interface, and the call is a value expression of type `type`.
-   If the callee is a function value, the call is an initializing expression
    whose type is the substituted return type of the function. When evaluated,
    the call expression will invoke the function and produce whatever value it
    returns.
-   If the callee is a bound method value, it behaves the same as a function
    value, except that the `self` parameter of the called function is bound to
    the `self` value in the bound method value.

### Generic callable parameters

A generic parameter can be constrained to be a callable type using the `Call`
interface:

```
interface Call(Args: type) {

>   let Result:! type;

Reviewer 4: Why are the argument types parameters but the result type an associated type? I can imagine there being good reasons for this, but would be helpful to articulate them here I think.

Author: Added an explanation.

  fn Op[self: Self](args: Args) -> Result;
}

> ```

Reviewer 3: I feel like we need 2-3 variations of this interface with different `self` parameters, with an `extend` relationship between them.

Author: Yeah, I think that's a good point, that we shouldn't forget about. I've added a section to "Future work" on this; I think we can decouple that from this proposal and add the extra interfaces separately.

        callable objects.
-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
    -   Function types in C and C++ are notoriously hard to read. We avoid this
        problem by not having signature-based function types.
-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
    -   As [future work](#function-pointers) we have the basis for a design for
        function pointers. C++ function pointers can be modeled in Carbon code
        as values that implement `Call`, as can member function pointers.

## Alternatives considered

### Signature-based function types

We could give each function signature a distinct type, as is done in C and C++.
This would provide a story for [function pointers](#function-pointers) that does
not rely on generics or type erasure or fancy representation optimizations.

The main down side of this approach is one we see frequently in C++: passing a
function to a template generic would be less efficient than passing a function

> object to the same template. For example, in C++, given

Reviewer 4: If this were solved in some other way, would that change the direction you'd want here?

I see other ways of avoiding this specific issue, so I'm curious to know how important this point of rationale ended up being.

Author: I think the point mentioned at the end of this section would still be relevant: giving functions unique types means that functions, lambdas, and function-like objects are more similar and have more uniform behavior. If the approach we took to avoid this specific issue affected the other cases in the same way, I think that would fully address the concern.

Depending on how we approach overloading, overload selection might be done by
checking each candidate in turn until we find one that matches, or checking them
all and somehow selecting the best match, and our current intent is to use the
former approach. Translated into this proposal, one important consequence is
that the function type corresponding to the set of overloaded functions has
multiple parameterized implementations of the `Call(...)` interface, and the
implementation selected for a particular call will need to follow whatever rules
the overloading mechanism picks. For example, if function overloading picks the
first matching function, then given the following, using placeholder syntax:

```
overloaded fn Abs[T:! Unsigned](x: T) -> T { return x; }
overloaded fn Abs[T:! Floating](x: T) -> T { return x < 0 ? -x : x }
```

we would synthesize implementations that behave as follows:

```
match_first {
  impl forall [T:! Unsigned] Abs as Call((T,)) where .Result = T { ... }
  impl forall [T:! Floating] Abs as Call((T,)) where .Result = T { ... }
}
```

For a type that is `Unsigned & Floating`, the former `impl` will be selected,
matching the behavior of overload selection.

> ### Overloading on expression category

Reviewer 3: It would be good to add that we want to, as future work, possibly overload on expression category for other parameters than the receiver.

It would also be good to have a future work section saying we will consider overloading on expression phase.

Author: Done.

    separated by commas, and if the argument list is not empty, an optional
    trailing comma is permitted but not required after the final argument.

Call syntax is syntactically equivalent to a primary expression followed by a
tuple literal, except that a tuple literal requires a trailing comma to form a
single-element tuple `(b,)`, whereas in call syntax both `a(b)` and `a(b,)` are
permitted.

### Direct calls

A call expression is a _direct call_ when the callee:

-   is the name of a parameterized entity, like a generic class or interface, or
-   has a function type or bound method type.

In a direct call, a call signature is available which is used to check the given
arguments against the callee's declared implicit and explicit parameters. This
checking proceeds as follows:

-   Argument deduction is performed by comparing the declared parameter types
    against the actual argument types and deducing values for implicit arguments
    that make the types equal.
-   Then, for each binding in the explicit parameter list in turn, all argument
    values that have been deduced are substituted into the parameter.

>     -   If the parameter is a `template :!` binding, the argument expression is
>         converted to a constant value phase expression of the type of the
>         binding.

Reviewer 4: I think missing a "template" here, and some expression phrase adjustments?

```suggestion
    -   If the parameter is a `template :!` binding, the argument expression is
        converted to a template constant expression phase of the type of the
        binding.
```

Author: Done.

    -   If the parameter is a checked `:!` binding, the argument expression is
        converted to a symbolic value phase expression of the type of the

>         binding.

Reviewer 4: And here missing "constant", and maybe restructuring a bit to orient around expression phases?

I'm also happy with both omitting "constant" if you'd prefer, but good to be consistent with the recent terminology choices.

```suggestion
    -   If the parameter is a checked `:!` binding, the argument expression is
        converted to a symbolic constant expression phase of the type of the
        binding.
```

Author: Done.

matching the behavior of overload selection.

### Overloading on expression category and phase

We should consider whether calls can be overloaded on the expression category
and expression phase of the callee and arguments. For compile-time arguments, we
may also wish to support overloading on the constant value.

Other languages support overloading on the expression category of the callable
object. Rust provides separate `Fn`, `FnMut`, and `FnOnce` to distinguish
between different ways of passing the callable object into an overloaded
function call, and C++ permits `operator()` to take `*this` as `const`,
non-`const`, or even `&&`.

In Carbon, we could follow the Rust approach and provide multiple `Call`
interfaces to handle different kinds of `self` parameter, modeled in a similar
way to the `IndexWith` and `IndirectIndexWith` interfaces used for subscripting.
A more general, ambitious and ergonomic approach that we have started exploring
would be to allow the function call signature to be described as part of the
call interface:

```carbon
impl MyCallable as call(T:! type, x: T) -> T;
```

This might be a shorthand syntax for some way of expressing the signature as an

> interface, or a new form of first-class language primitive.

Reviewer 4: Is there a good place to link to the issue where we're tracking the longer-term effort here?

Author: Done.

### Direct calls

> When the callee in a call expression is either the name of a parameterized
> entity like a generic class or interface or has a function type or bound method
> type, the call is a _direct call_.

Reviewer 3: I found this a bit hard to parse since it has the word "or" three times, with different precedences. I think you mean:

> When the callee in a call expression is either (the name of a parameterized entity like a generic (class or interface)) or (has a (function type or bound method type)), the call is a _direct call_.

I think this may just be a bit too complicated of a sentence structure. Maybe use bullets to make this clearer?

```suggestion
A call expression is a _direct call_ when the callee is:

-    the name of a parameterized entity, like a generic class or interface,
-    has a function type, or
-    has a bound method type.
```

or even

```suggestion
A call expression is a _direct call_ when the callee is:

-    the name of a parameterized entity, like a generic class or interface,
-    has a function type or bound method type.
```

    values that have been deduced are substituted into the parameter.

    -   If the parameter is a `template :!` binding, the argument expression is
        converted to have the same type as the binding and template constant
        expression phase.

>     -   If the parameter is a checked `:!` binding, the argument expression is

Reviewer 3: ```suggestion
    -   If the parameter is a symbolic `:!` binding, the argument expression is
```

This might be a shorthand syntax for some way of expressing the signature as an
interface, or a new form of first-class language primitive.

The options in this space are currently being explored in issue

> [#3154](https://github.com/carbon-language/carbon-lang/issues/3154)

Reviewer 3: ```suggestion
[#3154](https://github.com/carbon-language/carbon-lang/issues/3154).
```
